src/_pytest/recwarn.py

```diff
@@ -302,6 +303,17 @@ def __exit__(
 
         __tracebackhide__ = True
 
+        # BaseExceptions like pytest.{skip,fail,xfail,exit} or Ctrl-C within
+        # pytest.warns should *not* trigger "DID NOT WARN" and get suppressed
+        # when the warning doesn't happen. Control-flow exceptions should always
+        # propagate.
```

Comment:
Well, technically I see some packages (trio) using `with pytest.raises(GeneratorExit)` but it's a corner case, so I'm not going to argue.